### PR TITLE
Fix a bug where designer crash in RS4

### PR DESCRIPTION
**Pull request by ocalvo**
Issue: Bug 19752583: We should prevent UwpUIElementExtensions from being instantiated on RS4 machine

## PR Type
- Bugfix

## What is the current behavior?

Designer crashes when opening a form with a toolkit hosting control in RS4 machine.
In a RS4 machine, the type WindowsXamlManger does not exist and Xaml cannot be initialized in a Win32 process.
When run under the designer, this check will prevent an null reference exception happening in Windows.UI.Xaml.dll

## What is the new behavior?
Designer ignores toolkit hosting control in RS4 machine and prevents a null deref.

## PR Checklist

- [x] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: 
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [x] Contains **NO** breaking changes

## Other information
None

### CODE REVIEW

File: Microsoft.Toolkit.Wpf.UI.XamlHost/WindowsXamlHostWrapperConverter.cs [hunk 1]
```diff
@@ -2,39 +2,39 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
```
Comment (stevenbrix) [inline Microsoft.Toolkit.Wpf.UI.XamlHost/WindowsXamlHostWrapperConverter.cs]: looks like this is only whitespace changes, so it should probably be reverted (same with OSVersionHelper.cs)

File: Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs [hunk 2]
```diff
@@ -2,24 +2,49 @@
...
+using System.Security;
+
+namespace Microsoft.Toolkit.Wpf.UI.XamlHost
+{
+    /// <summary>
+    /// Extensions for use with UWP UIElement objects wrapped by the WindowsXamlHostBaseExt
+    /// </summary>
+    public static class UwpUIElementExtensions
+    {
+        [SecurityCritical]
```
Comment (stevenbrix) [inline Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs]: is SecurityCritical attribute necessary?

**Comment by stevenbrix** [comment]
Can you add some more information about what the crash is? Just from looking at this pr it's unclear what the issue is and how this fixes it

File: Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs (hunk 2 above)
Comment (Austin-Lamb) [reply Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs]: I don't think it should be.

Comment (drborts) [reply Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs]: I don't think it should be?

---
In reply to: [258750669](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/81#discussion_r258750669) [](ancestors = 258750669)

**Comment by ocalvo** [comment]
Sure

---
In reply to: [465828211](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/81#issuecomment-465828211) [](ancestors = 465828211)

File: Microsoft.Toolkit.Wpf.UI.XamlHost/WindowsXamlHostWrapperConverter.cs (hunk 1 above)
Comment (ocalvo) [reply Microsoft.Toolkit.Wpf.UI.XamlHost/WindowsXamlHostWrapperConverter.cs]: Its CR/LF changes, the files in git are using UNix style and VS is transform to Windows style.

---
In reply to: [258750369](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/81#discussion_r258750369) [](ancestors = 258750369)

Comment (stevenbrix) [reply Microsoft.Toolkit.Wpf.UI.XamlHost/WindowsXamlHostWrapperConverter.cs]: yeah, it's an unfortunate world we live in :/

**Comment by drborts** [comment]
test comment

**Comment by Austin-Lamb** [comment]
test comment

File: Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs [hunk 3]
```diff
@@ -2,24 +2,47 @@
...
+using System;
+
+namespace Microsoft.Toolkit.Wpf.UI.XamlHost
+{
+    /// <summary>
+    /// Extensions for use with UWP UIElement objects wrapped by the WindowsXamlHostBaseExt
+    /// </summary>
+    public static class UwpUIElementExtensions
+    {
+        private static bool IsDesktopWindowsXamlSourcePresent() => Windows.Foundation.Metadata.ApiInformation.IsApiContractPresent("Windows.UI.Xaml.Hosting.HostingContract", 3);
```
Comment (Austin-Lamb) [inline Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs]: >IsApiContractPresent [](start = 110, length = 20)

This call can be kinda slow, it'd be nice to cache the result.

File: Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs [hunk 4]
```diff
@@ -2,24 +2,47 @@
...
+        {
+            get
+            {
+                if (IsDesktopWindowsXamlSourcePresent())
+                {
+                    var result = Windows.UI.Xaml.DependencyProperty.RegisterAttached("Wrapper", typeof(System.Windows.UIElement), typeof(UwpUIElementExtensions), new Windows.UI.Xaml.PropertyMetadata(null));
+                    return result;
+                }
+
+                throw new NotImplementedException();
```
Comment (Austin-Lamb) [inline Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs]: >NotImplementedException [](start = 26, length = 23)

Where does this exception get caught (so the designer still functions)?

Comment (ocalvo) [reply Microsoft.Toolkit.Wpf.UI.XamlHost/UwpUIElementExtensions.cs]: Yes.

---
In reply to: [259030180](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/81#discussion_r259030180) [](ancestors = 259030180)

**Pull request merged**